### PR TITLE
Add OmnigenityHarmonics Objective and new OmnigenousField types

### LINKED ISSUE #1846: Add OmnigenityHarmonics Objective and new OmnigenousField types
# Description
This pull request introduces the OOPS (Omnigenity OPtimization like quasi-Symmetry) method into DESC, as described in [1]. OOPS leverages the concept of a homeomorphism to enable the optimization of omnigenity, as well as other concepts like pseudo-symetry and piecewise omnigenity, in spectral space, analogous to the optimization of quasisymmetry. A key advantage of this method is that it is independent of any specific parameterization. One can use the Cary-Shasharina mapping [2], the Landreman-Catto mapping [3], the current DESC mapping, or any others. It only requires a method for constructing the target magnetic field. Generally, it has better roubustness. The goal of this PR is to implement OOPS within DESC's flexible and efficient framework.
To facilitate discussion, I have created this **Draft Pull Request**. #1845 
# Core Changes
The main changes included in this PR are:
- New `OmnigenousField` Types: Two new parameterizations for omnigenous fields have been added in `magnetic_field/_core.py`:
  - `OmnigenousFieldOOPS`: The example mapping used in the OOPS paper [1].
  - `OmnigenousFieldLCForm`: The mapping form proposed by Landreman et al. [3], referred to here as the "LCForm".
- New `OmnigenityHarmonics` Objective: A new objective, `OmnigenityHarmonics`, has been added to `objectives/_omnigenity.py`. This objective enables the direct optimization of `OmnigenousField`, `OmnigenousFieldOOPS`, and `OmnigenousFieldLCForm` in spectral space.
- New Computations: Necessary parameters have been registered in `compute/_omnigenity.py`. The core addition is the function` _B_omni_nonsymmetric`, which calculates the non-symmetric components of the omnigenous field, `B(η,α)`.
- New Plotting Functions: To support visualization of these new field types, `plot_boozer_surface_XXX` routines for `OmnigenousFieldOOPS` and `OmnigenousFieldLCForm` have been implemented in `plotting.py`.
# Examples
Here are some examples in Ref. [1]. They can be optimized using DESC. We have implemented an optimization workflow in DESC with objectives, procedures, and resolutions that are nearly identical to the corresponding SIMSOPT scripts. Following the same process, starting from a circular torus, we use a resolution continuation method to optimize the configuration up to a spectral resolution of `M=N=7`. Depending on whether jax-cache is enabled, a precisely poloidal omnigenous (PO) configuration can be optimized within 10-15 minutes on a single NVIDIA RTX 5090 GPU. (With a configuration nearly identical to the one described in Ref. [1])
Image
# Known I/O Limitation
  While the I/O interface can successfully save and load the degrees of freedom of `OmnigenousFieldLCForm` (`s_dofs`, `d_dofs`), it does not currently support serializing the function definitions themselves. This means that when loading from a saved state, the user must manually re-supply the Python functions for `S_function` and `D_function`.
  
# References
[1] H. Liu, G. Yu, et al., "Optimizing omnigenity like quasisymmetry for stellarators," arxiv:2502.09350 (2025).
[2] J. R. Cary and S. G. Shasharina, Helical Plasma Confinement Devices with Good Confinement Properties, Phys. Rev. Lett. **78**, 674 (1997).
[3] M. Landreman and P. J. Catto, "Omnigenity as generalized quasisymmetry," Physics of Plasmas 19, 056103 (2012).

### PR DESCRIPTION
**Pull request by DMCXE**
# Description
This pull request introduces the OOPS (Omnigenity OPtimization like quasi-Symmetry) method into DESC, as described in [1]. OOPS leverages the concept of a homeomorphism to enable the optimization of omnigenity, as well as other concepts like pseudo-symetry and piecewise omnigenity, in spectral space, analogous to the optimization of quasisymmetry. A key advantage of this method is that it is independent of any specific parameterization. One can use the Cary-Shasharina mapping [2], the Landreman-Catto mapping [3], the current DESC mapping, or any others. It only requires a method for constructing the target magnetic field. Generally, it has better roubustness. The goal of this PR is to implement OOPS within DESC's flexible and efficient framework.
To facilitate discussion, I have created this **Draft Pull Request**. See same issue in #1846 
# Core Changes
The main changes included in this PR are:
- New `OmnigenousField` Types: Two new parameterizations for omnigenous fields have been added in `magnetic_field/_core.py`:
  - `OmnigenousFieldOOPS`: The example mapping used in the OOPS paper [1].
  - `OmnigenousFieldLCForm`: The mapping form proposed by Landreman et al. [3], referred to here as the "LCForm".
- New `OmnigenityHarmonics` Objective: A new objective, `OmnigenityHarmonics`, has been added to `objectives/_omnigenity.py`. This objective enables the direct optimization of `OmnigenousField`, `OmnigenousFieldOOPS`, and `OmnigenousFieldLCForm` in spectral space.
- New Computations: Necessary parameters have been registered in `compute/_omnigenity.py`. The core addition is the function` _B_omni_nonsymmetric`, which calculates the non-symmetric components of the omnigenous field, `B(η,α)`.
- New Plotting Functions: To support visualization of these new field types, `plot_boozer_surface_XXX` routines for `OmnigenousFieldOOPS` and `OmnigenousFieldLCForm` have been implemented in `plotting.py`.
# Examples
Here are some examples in Ref. [1]. They can be optimized using DESC. We have implemented an optimization workflow in DESC with objectives, procedures, and resolutions that are nearly identical to the corresponding SIMSOPT scripts. Following the same process, starting from a circular torus, we use a resolution continuation method to optimize the configuration up to a spectral resolution of `M=N=7`. Depending on whether jax-cache is enabled, a precisely poloidal omnigenous (PO) configuration can be optimized within 10-15 minutes on a single NVIDIA RTX 5090 GPU. (With a configuration nearly identical to the one described in Ref. [1])
1280X1280
# Known I/O Limitation
  While the I/O interface can successfully save and load the degrees of freedom of `OmnigenousFieldLCForm` (`s_dofs`, `d_dofs`), it does not currently support serializing the function definitions themselves. This means that when loading from a saved state, the user must manually re-supply the Python functions for `S_function` and `D_function`.
  
# References
[1] H. Liu, G. Yu, et al., "Optimizing omnigenity like quasisymmetry for stellarators," arxiv:2502.09350 (2025).
[2] J. R. Cary and S. G. Shasharina, Helical Plasma Confinement Devices with Good Confinement Properties, Phys. Rev. Lett. **78**, 674 (1997).
[3] M. Landreman and P. J. Catto, "Omnigenity as generalized quasisymmetry," Physics of Plasmas 19, 056103 (2012).

### CODE REVIEW

File: desc/compute/_omnigenity.py [hunk 1]
```diff
@@ -938,3 +939,509 @@ def _isodynamicity(params, transforms, profiles, data, **kwargs):
...
+    coordinates="rtz",
+    data=["eta_OOPS"],
+    parameterization="desc.magnetic_fields._core.OmnigenousFieldOOPS",
+)
+def _B_omni_OOPS(params, transforms, profiles, data, **kwargs):
+    def fake_B_target(eta):
+        return 1 + 0.25 + 0.25 * jnp.cos(eta - jnp.pi)
+
+    eta = transforms["grid"].meshgrid_reshape(data["eta_OOPS"], "rtz")
+    B = fake_B_target(eta)
```
Comment (dpanici) [inline desc/compute/_omnigenity.py:1250]: what is this fake_B_target used for?

File: desc/objectives/_omnigenity.py [hunk 2]
```diff
@@ -976,3 +981,513 @@ def compute(self, params, constants=None):
...
+                    self._field_data_keys,
+                    params=self._field.params_dict,
+                    transforms=self._constants["field_transforms"],
+                    profiles={},
+                    helicity=self._constants["helicity"],
+                )
+            elif self._field_type == "oops":
+                # precompute the field data since it is fixed during the optimization
+                field_data = compute_fun(
+                    "desc.magnetic_fields._core.OmnigenousFieldOOPS",
```
Comment (dpanici) [inline desc/objectives/_omnigenity.py]: you can pass in the `self._field` object here instead of the string parametrization, so you dont need all of these if branches

File: desc/compute/_omnigenity.py [hunk 3]
```diff
@@ -938,3 +939,509 @@ def _isodynamicity(params, transforms, profiles, data, **kwargs):
...
+    eta = transforms["grid"].meshgrid_reshape(data["eta_LCForm"], "rtz")
+    B = fake_B_target(eta)
+    # Here B is 2d
+    B = jnp.moveaxis(B, 0, 1)
+    data["|B|_LCForm"] = B.flatten(order="F")
+    return data
+
+
+def _B_omni_nonsymmetric(
+    B_eta_alpha, M_harmonics, N_harmonics, field_grid, is_imag=True
```
Comment (dpanici) [inline desc/compute/_omnigenity.py]: is it possible to use the DESC transform API do do this fitting? instead of implementing your own FFT here

Comment (dpanici) [reply desc/compute/_omnigenity.py]: you would then only need to pass in the transform object, instead of M, N and field_grid

File: desc/plotting.py [hunk 4]
```diff
@@ -4224,3 +4226,297 @@ def plot_logo(save_path=None, **kwargs):
         fig.savefig(save_path, facecolor=fig.get_facecolor(), edgecolor="none")
 
     return fig, ax
+
+
+def plot_boozer_surface_OOPS(
```
Comment (dpanici) [inline desc/plotting.py]: if possible, could these be added as options to the usual `plot_boozer_surface`, like how it is currently setup to handle the DESC `OmnigenousField` class to plot |B|? I think that would be doable and would reduce the amount of plotting functions we have to maintain

**Comment by rahulgaur104** [comment]
Another big, code-related change would be to create a PR from a branch on the DESC repo, not on a fork from your repo.
I don't remember how but it helps with tests and merging into master.

File: desc/compute/_omnigenity.py [hunk 5]
```diff
@@ -938,3 +939,509 @@ def _isodynamicity(params, transforms, profiles, data, **kwargs):
...
+    It is used in OOPS to define the omnigenity symmetry angle.
+    """
+    if "S_list" not in params:
+        raise ValueError("S_list parameter is required for Landreman Form")
+    data["S_list"] = jnp.array(params["S_list"])
+    return data
+
+
+@register_compute_fun(
+    name="D_list",
```
Comment (rahulgaur104) [inline desc/compute/_omnigenity.py]: D_list has been defined twice? Is this the same variable?

File: desc/compute/_omnigenity.py [hunk 6]
```diff
@@ -938,3 +939,509 @@ def _isodynamicity(params, transforms, profiles, data, **kwargs):
...
+    eta = transforms["grid"].meshgrid_reshape(data["eta_OOPS"], "rtz")
+    B = fake_B_target(eta)
+    # Here B is 2d
+    B = jnp.moveaxis(B, 0, 1)
+    data["|B|_OOPS"] = B.flatten(order="F")
+    return data
+
+
+@register_compute_fun(
+    name="S_list",
```
Comment (rahulgaur104) [inline desc/compute/_omnigenity.py]: S_list has been defined twice? Is this the same variable?

**Review by rahulgaur104 — CHANGES_REQUESTED** [review]

File: desc/compute/_omnigenity.py (hunk 5 above)
Comment (rahulgaur104) [reply desc/compute/_omnigenity.py]: Seems like the form is different. The second D_list is Landreman form so the name must be changed.

File: desc/compute/_omnigenity.py (hunk 3 above)
Comment (dpanici) [reply desc/compute/_omnigenity.py]: @dpanici share code snippet

Comment (dpanici) [reply desc/compute/_omnigenity.py]: Probably longer than necessary code snippet but basically this is an example of fitting some data in DESC using transforms. In your case you can just pre-build the transform either in the objective function `.build` method, or in the `get_transforms` function (sort of like how we already do for some of our[ Boozer transforms](https://github.com/PlasmaControl/DESC/blob/c704244365cd99958b4493694e8ca2672d8a9894/desc/compute/_omnigenity.py#L76) or more similarly like how we do for [existing omnigenitu](https://github.com/PlasmaControl/DESC/blob/c704244365cd99958b4493694e8ca2672d8a9894/desc/compute/utils.py#L627)). You would just need to add something like "B" or even "h" if it is directly applicable, to the `transforms` part of whatever compute method you have. 

Let us know if you need any help as we know the structure can be a bit confusion

```python
from desc.examples import get
from desc.grid import LinearGrid
from desc.basis import DoubleFourierSeries
from desc.transform import Transform

eq = get("ESTELL")

M_basis = 20
N_basis=20

basis = DoubleFourierSeries(M=M_basis,
                            N=N_basis,
                            NFP=eq.NFP
                            ,sym=False) # can be "cos", "sin", "False"


grid = LinearGrid(rho=1.0, M=2*M_basis, N=2*N_basis, NFP=eq.NFP)

# this essentailly contains the vandermonde matrices for
# the given basis at the given grid nodes i.e.
# the matrix A s.t. A @ x_mn = x(theta, zeta)
transform = Transform(basis=basis,
                      grid=grid,
                      build_pinv=True) # make sure this is True so pseudoinverse is built
# here I will fit |B| to 
# the computational theta and zeta 
# first compute |B| at the grid nodes
B = eq.compute("|B|",grid=grid)["|B|"]

# next, use transform.fit to find B_mn
B_mn = transform.fit(B)

# then I will re-evaluate on a finer grid and plot for comparison
grid_dense  = LinearGrid(rho=1.0, M=3*M_basis, N=3*N_basis, NFP=eq.NFP)

transform_dense = Transform(grid=grid_dense, basis=basis)

B_orig = transform.transform(B_mn)
B_dense = transform_dense.transform(B_mn)

# plot just to visually show this works
import matplotlib.pyplot as plt
plt.figure()
theta = grid.nodes[:,1].reshape((grid.num_zeta, grid.num_theta))
zeta = grid.nodes[:,2].reshape((grid.num_zeta, grid.num_theta))
plt.contour(zeta, theta, B_orig.reshape((grid.num_zeta, grid.num_theta)), levels=20)

plt.figure()
theta = grid_dense.nodes[:,1].reshape((grid_dense.num_zeta, grid_dense.num_theta))
zeta = grid_dense.nodes[:,2].reshape((grid_dense.num_zeta, grid_dense.num_theta))
plt.contour(zeta, theta, B_dense.reshape((grid_dense.num_zeta, grid_dense.num_theta)), levels=20)
```
